modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/request/GsonRequest.mustache

```diff
@@ -0,0 +1,119 @@
+package {{packageName}}.request
+
+import com.android.volley.NetworkResponse
+import com.android.volley.ParseError
+import com.android.volley.Request
+import com.android.volley.Response
+import com.android.volley.toolbox.HttpHeaderParser
+import com.google.gson.Gson
+import com.google.gson.GsonBuilder
+import com.google.gson.JsonSyntaxException
+import java.io.UnsupportedEncodingException
+import java.nio.charset.Charset
+import java.net.HttpURLConnection
+import java.lang.reflect.Type
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.time.OffsetDateTime
+
+import {{packageName}}.infrastructure.OffsetDateTimeAdapter
+import {{packageName}}.infrastructure.LocalDateTimeAdapter
+import {{packageName}}.infrastructure.LocalDateAdapter
+import {{packageName}}.infrastructure.ByteArrayAdapter
+
+class GsonRequest<T>(
+    method: Int,
+    url: String,
+    private val body: Any?,
+    private val headers: Map<String, String>?,
+    private val params: MutableMap<String, String>?,
+    private val contentTypeForBody: String?,
+    private val encodingForParams: String?,
+    private val gsonAdapters: Map<Type, Any>?,
+    private val type: Type,
+    private val listener: Response.Listener<T>,
+    errorListener: Response.ErrorListener
+) : Request<T>(method, url, errorListener) {
+
+    val gsonBuilder: GsonBuilder = GsonBuilder()
+        .registerTypeAdapter(OffsetDateTime::class.java, OffsetDateTimeAdapter())
+        .registerTypeAdapter(LocalDateTime::class.java, LocalDateTimeAdapter())
+        .registerTypeAdapter(LocalDate::class.java, LocalDateAdapter())
+        .registerTypeAdapter(ByteArray::class.java, ByteArrayAdapter())
+        .apply {
+            gsonAdapters?.forEach {
+                this.registerTypeAdapter(it.key, it.value)
+            }
+        }
+
+    val gson: Gson by lazy {
+        gsonBuilder.create()
+    }
```

Comment:
@alisters why don't we use the Serializer.gson instead of creating a new Gson here?

Reply:
The consumer of the this, the builder of the gson request, needs to pass custom type adapters. And the types that respective builders know about are limited to that builder - so a singleton gson adapter wouldn't work. See code example below - In this case the payeeTypeAdapter is used to cope with sub typing. Thats the design rationale. That said, maybe there is a better way to do this and default to Serializer.gson if no gsonAdapters are provided ? I'm thinking about it :-)

#####
 @Provides
    internal fun providePayeeApi(
        context: Context,
        restService: IRestService,
        configurationService: IConfigurationService,
        sessionService: ISessionService
    ): PayeeApi {
        val payeeAdapter = mapOf(Pair<Type, Object>(Payee::class.java, PayeeTypeAdapter() as Object))
        return PayeeApi(
            context = context,
            requestQueue = restService.getRequestQueue(),
            requestFactory = RequestFactory(
                headerFactories = listOf(createSessionHeaderFactory(sessionService, restService), createTraceHeaderFactory()),
                postProcessors = listOf(retryPolicySetter),
                gsonAdapters = payeeAdapter
            ),
            basePath = configurationService.getBaseUrl()
        )
    }

Reply:
Thanks for the explanation, some of my questions are because the volley integration is very different from the others in the kotlin generator

Reply:
@4brunu - general comments, no reply necessary. 

The other point i'd look at for these serialisers is the thread safety. I'm not sure if they are thread safe, and volley, especially with our non blocking co-routines, can definitely dispatch multiple calls in parallel. 

It is different, and TBH we couldn't decide which generator to use to start with, or whether to create a new one. To add a bit of the back story - In the related issue/feature request i mentioned there was 
"A volley client library in the android generator, which generates java.
A kotlin client generator with various libraries but not Volley.
Searching the issues shows the volley client may need attention - #9293 "

So the major reasons to put it in the Kotlin client 
- were to reuse the kotlin base types. 
- to make use of coroutines, since we were using them for room etc. 

We also looked into porting our code from volley to retrofit, as we wanted to use co-routines, and the kotlin retrofit client supports those (https://openapi-generator.tech/docs/generators/kotlin/) , but on looking at the code for those it was too far away from our manual volley code that we already had.

Reply:
By the way, just curious, but why do you prefer Volley instead of OkHttp or Retrofit?

Reply:
I wouldn't say i have a preference really - i've never used the others. In kotlin i've mostly done server side stuff - using spring and web client recently. We are retrofitting a code gen/contract first approach to a set of existing clients. The android client already used volley. The author of that chose volley because the android doco seemed to default to using it, especially for kotlin based android apps. That said I like it - the design allowed mixing in various functionality easily using dependency injection with Dagger - so fairly painless. 

BTW if you haven't seen it from already the README does cover some of our design - and how we've tried to make the generator surface Volleys functionality. We don't use that much of it, but i tried to make the generator generic enough that other people could use and extend it for their use cases.